### ISSUE TITLE
Wrong formula's class notation

### ISSUE THREAD

**Issue by tommasongr** [dart-lang/homebrew-dart#170]
Hi there,

I'm opening this because a user of our Homebrew client ([brewerx.app](https://brewerx.app)) has reported an issue with this repository.

We discovered that all the formulas after dart@2.9 are not using the expected class name notation. E.g: `DartAT3_2_4` instead of `DartAT324`.

Even with vanilla Homebrew we got the error:
```shell
$ brew install dart@3.2

Error: No available formula with the name "dart@3.2.4". Did you mean dart@3.2.1, dart@3.1.4, dart@3.0.4, dart@3.2.3, dart@3.2.2, dart@3.2.0, dart@3.1.2, dart@3.0.2, dart@2.3, dart@2.4, dart@3.0.6, dart@3.0.5, dart@3.0.3, dart@3.0.1, dart@3.0.0, dart@3.1.1, dart@3.1.0, dart@3.0.7, dart@3.1.5, dart@3.1.3, dart@2.14, dart@2.9, dart@2.7, dart@2.6, dart@2.5, dart@2.2, dart@2.1, dart@2.0 or dart@2.8?
In formula file: /opt/homebrew/Library/Taps/dart-lang/homebrew-dart/Formula/dart@3.2.4.rb
Expected to find class DartAT324, but only found: DartAT3_2_4.
```

If you weren't aware of the issue, please consider to fix the malformed files.

Thanks

**Comment by sortie** [comment]
Hi @tommasongr, I was actually worried about this, but didn't have a mac handy to test on. Thanks for helping me out, I don't know so much about homebrew. I'll cook up a solution :)

The big problem here is that the homebrew notation is ambiguous. E.g. the minor and patch release digits may get larger than 9. For instance, does DartAT4321 mean Dart 4.32.1 or 4.3.21? Is there a homebrew solution for this problem? Do you know more? I suppose it's not so bad since at least it's in different files.

**Comment by tommasongr** [comment]
Hi @sortie, thanks for the fast reply! I'm happy to give you a hand and don't worry... these kind of issues are more common than you might think, unfortunately for us hahah

Anyway, I have to admit that I'm not super familiar with the Homebrew internals. We're mostly developing the macOS client. However I have a theory but take it with a pinch of salt.

I'm familiar with Rails and ruby based applications like Homebrew. They almost all use, including Homebrew for what I've seen, a code loading library called [Zeitwerk](https://github.com/fxn/zeitwerk). The notation you're not respecting comes from Zeitwerk, not from Homebrew. In Zeitwerk the source of truth is the file name, not the name of the class. That is inferred. So if you call your file `dart@3.2.4` you have to call the class `DartAT324` respectively. Don't worry about the actual versioning, the class name has nothing to do with that.

I hope this is useful.

**Comment by sortie** [comment]
Thanks @tommasongr for all your assistance and reporting the issue. I have a fix up in https://dart-review.googlesource.com/c/homebrew-dart/+/345700

Sorry about the breakage, I'm a Linux dev and recently switched away from my old macOS laptop, so I didn't have a quick way to test locally and we don't have any test coverage for these versioned formulas, so it slipped through. Hopefully everything will be working perfectly now because my goal was to make the homebrew experience much better by automatically generating these formulas and making every stable release available immediately for you :)

**Comment by sortie** [comment]
This should be fixed in 2dbf0034161b137f00df779bc21e769ff5f51ae5.
